File: 2024/Python/Day01/Day01.py
from typing import TextIO


def get_file_data(file_name) -> list[str]:
    f: TextIO = open(file_name)
    data: list[str] = []
    for line in f:
        data.append(line.rstrip())
    return data


def part_one():
    result = 0
    left = []
    right = []
    for i in file_data:
        left.append(int(i.split(" ")[0]))
        right.append(int(i.split(" ")[-1]))
    for index, i in enumerate(sorted(left)):
        result += abs(sorted(left)[index] - sorted(right)[index])
    print(f"part one: {result}")


def part_two():
    result = 0
    left = []
    right = []
    for i in file_data:
        left.append(int(i.split(" ")[0]))
        right.append(int(i.split(" ")[-1]))
    for i in left:
        result += i * right.count(i)
    print(f"part two: {result}")


file_data = get_file_data("input")
part_one()
part_two()

File: 2024/Python/Day01/Day01.cpp
#include <vector>
#include <iostream>
#include <fstream>
#include <sstream>
#include <algorithm>
using namespace std;

int main() {
    ifstream file("input");
    if (!file.is_open()) {
        cerr << "Unable to open file" << endl;
        return 1;
    }
    int total;
    vector<int> left;
    vector<int> right;
    string line;
    // This is such a terrible way to split data but it works
    while (getline(file, line)) {
        int tmp;
        istringstream stream(line);
        bool first = true;
        while (stream >> tmp) {
            if (first) left.push_back(tmp);
            if (!first) right.push_back(tmp);
            first = !first;
        }
    }
    sort(left.begin(), left.end());
    sort(right.begin(), right.end());
    for (int i = 0; i < left.size(); i++) total += abs(left.at(i) - right.at(i));
    cout << "part one: " << total << endl;
    // whether vector is sorted or not doesn't matter; just use sorted vector
    total = 0;
    // nested for loop. takes an element from left, checks repeats on right, multiply repeat and left element and add to total
    for (int i : left) {
        total += i * count(right.begin(), right.end(), i);
    }
    cout << "part two: " << total;
    file.close();
    return 0;
}
